File: apps/mobile/lib/oidc.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:app_links/app_links.dart';
import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:url_launcher/url_launcher.dart';

import 'config.dart';
import 'debug_log.dart';

const oidcPkceVerifierKey = 'hris_oidc_pkce_verifier';
const oidcPkceStateKey = 'hris_oidc_pkce_state';

Future<void> clearOidcPrefs([SharedPreferences? prefs]) async {
  final p = prefs ?? await SharedPreferences.getInstance();
  await p.remove(oidcPkceVerifierKey);
  await p.remove(oidcPkceStateKey);
}

/// Authorization code + PKCE against Cognito Hosted UI (lab defaults) or Keycloak.
class OidcLogin {
  OidcLogin({http.Client? httpClient, AppLinks? appLinks})
      : _http = httpClient ?? http.Client(),
        _appLinks = appLinks ?? AppLinks();

  final http.Client _http;
  final AppLinks _appLinks;

  Future<String> signIn() async {
    if (!AppConfig.oidcConfigured) {
      throw StateError(
        'OIDC is not configured. Use Continue as employee, or pass '
        'OIDC_AUTHORIZE_URL / OIDC_TOKEN_URL dart-defines (see README).',
      );
    }

    final prefs = await SharedPreferences.getInstance();
    final verifier = randomUrlSafe(48);
    final state = randomUrlSafe(16);
    await prefs.setString(oidcPkceVerifierKey, verifier);
    await prefs.setString(oidcPkceStateKey, state);

    final authorize = Uri.parse(AppConfig.oidcAuthorizeUrl).replace(
      queryParameters: {
        'client_id': AppConfig.oidcClientId,
        'response_type': 'code',
        'redirect_uri': AppConfig.oidcRedirectUri,
        'scope': AppConfig.oidcScopes,
        'code_challenge': s256Challenge(verifier),
        'code_challenge_method': 'S256',
        'state': state,
      },
    );

    LabDebugLog.instance.oidc('authorize', AppConfig.oidcRedirectUri);

    final callback = Completer<Uri>();

    void onUri(Uri? uri) {
      if (uri == null || callback.isCompleted) return;
      if (!_isOAuthRedirect(uri)) return;

      final returnedState = uri.queryParameters['state'];
      if (returnedState != state) {
        LabDebugLog.instance.warn(
          'ignored stale callback state=$returnedState expected=$state',
        );
        return;
      }

      callback.complete(uri);
    }

    final sub = _appLinks.uriLinkStream.listen(onUri);

    try {
      // Drain cached deep link from a prior attempt before opening Hosted UI.
      onUri(await _appLinks.getInitialLink());

      final opened = await launchUrl(authorize, mode: LaunchMode.externalApplication);
      if (!opened) {
        throw StateError('Could not open the identity provider in a browser.');
      }
      LabDebugLog.instance.oidc('browser opened');
      LabDebugLog.instance.oidc('waiting for callback');

      final redirected = await callback.future.timeout(const Duration(minutes: 3));
      LabDebugLog.instance.oidc('callback', redirected.toString());

      final error = redirected.queryParameters['error'];
      if (error != null) {
        throw StateError(formatOidcRedirectError(error, redirected.queryParameters['error_description']));
      }
      final code = redirected.queryParameters['code'];
      if (code == null || code.isEmpty) {
        throw StateError('OIDC callback missing code');
      }
      final savedVerifier = prefs.getString(oidcPkceVerifierKey);
      if (savedVerifier == null || savedVerifier.isEmpty) {
        throw StateError('OIDC session expired. Try Sign in again.');
      }
      return _exchange(code: code, verifier: savedVerifier);
    } finally {
      await sub.cancel();
      await clearOidcPrefs(prefs);
    }
  }

  bool _isOAuthRedirect(Uri uri) {
    final expected = Uri.parse(AppConfig.oidcRedirectUri);
    if (uri.scheme != expected.scheme) return false;
    if (expected.host.isNotEmpty && uri.host != expected.host) return false;
    return uri.queryParameters.containsKey('code') ||
        uri.queryParameters.containsKey('error');
  }

  Future<String> _exchange({required String code, required String verifier}) async {
    LabDebugLog.instance.oidc('token exchange');
    final res = await _http.post(
      Uri.parse(AppConfig.oidcTokenUrl),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: {
        'grant_type': 'authorization_code',
        'client_id': AppConfig.oidcClientId,
        'code': code,
        'redirect_uri': AppConfig.oidcRedirectUri,
        'code_verifier': verifier,
      },
    );
    if (res.statusCode < 200 || res.statusCode >= 300) {
      LabDebugLog.instance.error('token exchange ${res.statusCode} ${res.body}');
      throw StateError('IdP token exchange failed (${res.statusCode}): ${res.body}');
    }
    LabDebugLog.instance.oidc('token exchange ok');
    final json = jsonDecode(res.body) as Map<String, dynamic>;
    // Nest JwtAuthGuard verifies via JWKS. Prefer id_token so email + groups map
    // onto the seeded lab user (same as the React portals).
    final token = (json['id_token'] as String?) ?? (json['access_token'] as String?);
    if (token == null || token.isEmpty) {
      throw StateError('IdP response missing id_token/access_token');
    }
    return token;
  }
}

String randomUrlSafe(int byteCount) {
  final rng = Random.secure();
  final bytes = List<int>.generate(byteCount, (_) => rng.nextInt(256));
  return base64UrlEncode(bytes).replaceAll('=', '');
}

String s256Challenge(String verifier) {
  final digest = sha256.convert(utf8.encode(verifier));
  return base64UrlEncode(digest.bytes).replaceAll('=', '');
}

/// User-facing message for OAuth error redirects (matching [state] only).
String formatOidcRedirectError(String error, String? description) {
  if (error == 'login_pages_unavailable') {
    return 'Cognito login pages are unavailable. In AWS Cognito, assign a managed '
        'login style to the hris-mobile app client (${AppConfig.oidcClientId}), '
        'then try Sign in again.';
  }
  return description ?? error;
}

File: apps/web-onboarding/src/oidc.ts
const VERIFIER_KEY = 'hris_pkce_verifier';
const STATE_KEY = 'hris_pkce_state';

function clientId(): string {
  return import.meta.env.VITE_OIDC_CLIENT_ID ?? '';
}

export function oidcConfigured(): boolean {
  return Boolean(import.meta.env.VITE_OIDC_AUTHORIZE_URL && import.meta.env.VITE_OIDC_TOKEN_URL && clientId());
}

export function isLocalViteHost(): boolean {
  const host = window.location.hostname;
  return host === 'localhost' || host === '127.0.0.1';
}

export function showDevBypass(): boolean {
  if (!isLocalViteHost()) return false;
  return import.meta.env.VITE_AUTH_DEV_BYPASS !== 'false';
}

function b64url(bytes: Uint8Array): string {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function randomUrlSafe(byteLen: number): string {
  return b64url(crypto.getRandomValues(new Uint8Array(byteLen)));
}

async function challengeS256(verifier: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verifier));
  return b64url(new Uint8Array(digest));
}

export function redirectUri(): string {
  return window.location.origin;
}

export async function beginHostedUiLogin(): Promise<void> {
  const authorize = import.meta.env.VITE_OIDC_AUTHORIZE_URL;
  const id = clientId();
  if (!authorize || !id) {
    throw new Error('Cognito Hosted UI is not configured (VITE_OIDC_AUTHORIZE_URL / VITE_OIDC_CLIENT_ID)');
  }
  const verifier = randomUrlSafe(32);
  const state = randomUrlSafe(16);
  sessionStorage.setItem(VERIFIER_KEY, verifier);
  sessionStorage.setItem(STATE_KEY, state);
  const params = new URLSearchParams({
    client_id: id,
    response_type: 'code',
    scope: 'openid email profile',
    redirect_uri: redirectUri(),
    state,
    code_challenge: await challengeS256(verifier),
    code_challenge_method: 'S256',
  });
  window.location.assign(`${authorize}?${params.toString()}`);
}

let consumeOnce: Promise<string | null> | undefined;

export function consumeAuthCodeIfPresent(): Promise<string | null> {
  if (!consumeOnce) consumeOnce = consumeAuthCode();
  return consumeOnce;
}

async function consumeAuthCode(): Promise<string | null> {
  const params = new URLSearchParams(window.location.search);
  const err = params.get('error');
  if (err) {
    const desc = params.get('error_description') ?? err;
    stripAuthParams();
    throw new Error(`Sign-in failed: ${desc}`);
  }
  const code = params.get('code');
  if (!code) return null;
  const state = params.get('state');
  const expected = sessionStorage.getItem(STATE_KEY);
  const verifier = sessionStorage.getItem(VERIFIER_KEY);
  sessionStorage.removeItem(STATE_KEY);
  sessionStorage.removeItem(VERIFIER_KEY);
  if (!state || state !== expected || !verifier) {
    stripAuthParams();
    throw new Error('Sign-in session expired. Try again.');
  }
  const tokenUrl = import.meta.env.VITE_OIDC_TOKEN_URL;
  const id = clientId();
  if (!tokenUrl || !id) throw new Error('VITE_OIDC_TOKEN_URL / VITE_OIDC_CLIENT_ID missing');
  const res = await fetch(tokenUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: id,
      code,
      redirect_uri: redirectUri(),
      code_verifier: verifier,
    }),
  });
  stripAuthParams();
  if (!res.ok) throw new Error(`Token exchange failed (${res.status})`);
  const json = (await res.json()) as { access_token?: string; id_token?: string };
  // Email-alias pool: access-token `username` is a UUID. ID token has email + groups.
  const jwt = json.id_token || json.access_token;
  if (!jwt) throw new Error('Token response missing JWT');
  return jwt;
}

function stripAuthParams() {
  const url = new URL(window.location.href);
  for (const key of ['code', 'state', 'error', 'error_description']) {
    url.searchParams.delete(key);
  }
  const search = url.searchParams.toString();
  window.history.replaceState({}, document.title, url.pathname + (search ? `?${search}` : '') + url.hash);
}

export function hostedUiLogout(): void {
  const logout = import.meta.env.VITE_OIDC_LOGOUT_URL;
  const id = clientId();
  if (!logout || !id) return;
  window.location.assign(
    `${logout}?${new URLSearchParams({ client_id: id, logout_uri: redirectUri() }).toString()}`,
  );
}
